File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.balanced = True
    def _get_height(self, node: Optional[TreeNode]) -> int:
        if not node:
            return 0
        left = self._get_height(node.left)
        right = self._get_height(node.right)
        if self.balanced:
            self.balanced = abs(left-right) <= 1
        return 1 + max(left, right)
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self._get_height(root)
        return self.balanced

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    void dfs(TreeNode* node, int k, vector<int>& vec){
        if (!node) return;

        dfs(node->left, k, vec);

        if (vec.size() < k){
            vec.push_back(node->val);
        }
        else return;

        dfs(node->right, k , vec);
    }
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> res;
        dfs(root, k, res);
        return res.back();
    }
};
